L58-python-async-api.md

```diff
@@ -0,0 +1,453 @@
+
+There are three mechanisms to schedule coroutines:
+
+1. Await the coroutine `await asyncio.sleep(1)`;
+2. Submit the coroutine to the event loop object `loop.call_soon(coro)`;
+3. Turn coroutine into an `asyncio.Task` object `asyncio.ensure_future(coro)`.
+
+## Proposal
+
+This gRFC intended to fully utilize `asyncio` and create a Pythonic paradigm for
+gRPC Python programming in Python 3 runtime.
+
+### Two Sets of API In One Package
+
+The new API will be isolated from the current API. The implementation of the new
+API is an entirely different stack than the current stack. On the downside, most
+gRPC objects and definitions can't be shared between these two sets of API.
+After all, it is not a good practice to introduce uncertainty (return coroutine
+or regular object) to our API. This decision is made to respect our contract of
+API since GA in 2016, including the underlying behaviors. **Developers who use
+our current API should not be affected by this effort.**
+
+For users who want to migrate to new API, the granularity migration is per
+channel, per server level. For both channel-side and server-side, the adoption
+of `asyncio` does not only include cosmetic changes of adding `async`/`await`
+keywords throughout the code base, it also requires thinking about the change
+from a potentially multi-threaded application to a single-threaded asynchronous
+application. For example, the synchronization methods have to be changed to use
+the `asyncio` edition (see [Synchronization
+Primitives](https://docs.python.org/3/library/asyncio-sync.html)).
+
+### Overview of Async API
+
+All `asyncio` related API will be kept inside the `grpc.aio` module. Take the
+most common use case as an example, the new API will diverge from the old API
+since the creation of `Channel` and `Server` object. New creation methods are
+added to instantiate async `Server` or `Channel` object:
+
+```Python
+import grpc
+server = grpc.aio.server(...)
+channel = grpc.aio.insecure_channel(...)
+channel = grpc.aio.secure_channel(...)
+channel = grpc.aio.intercept_channel(...)
+```
+
+To reduce cognitive burden, the new `asyncio` API should share the same
+parameters as the current API, and keep the usage similar, except the following
+cases:
+
+* `grpc.aio.server` no longer require the application to provide a
+  `ThreadPoolExecutor`;
+* Interfaces returning `Future` object are replaced.
+
+### Demo Snippet of Async API
+
+Server side:
+```Python
+class AsyncGreeter(helloworld_pb2_grpc.GreeterServicer):
+
+    async def SayHello(self, request, context):
+        await asyncio.sleep(1)
+        return helloworld_pb2.HelloReply(message='Hello, %s!' % request.name)
+
+server = grpc.aio.server()
+server.add_insecure_port(":50051")
+helloworld_pb2_grpc.add_GreeterServicer_to_server(Greeter(), server)
+server.start()
+await server.wait_for_termination()
+```
+
+Client side:
+```Python
+# Python 3.6 and lower
+with grpc.aio.insecure_channel("localhost:50051") as channel:
```

Comment:
Give support for the none synchronous and the asynchronous one might be a bit confusing for the users, since we are not gonna be able to implement everything for both use cases. I would get stick on only implementing the asynchronous one.

Reply:
Considering `async with` is supported since 3.5, I will remove the first use case.